tests/cecil-tests/GenericPInvokes.cs

```diff
@@ -0,0 +1,283 @@
+		}
+
+		static OpCode [] reasonableOps = new OpCode [] {
+			OpCodes.Ldarg_0, OpCodes.Ldarg_1, OpCodes.Ldarg_2,
+			OpCodes.Ldarg_3, OpCodes.Ldarg_S, OpCodes.Ldloc_0,
+			OpCodes.Ldloc_1, OpCodes.Ldloc_2, OpCodes.Ldloc_3,
+ 			OpCodes.Ldloc_S
+		};
+		static bool IsLastArgUsageIsReasonable (Instruction instr)
+		{
+			return Array.IndexOf (reasonableOps, instr.OpCode) >= 0;
+		}
+
+		static bool IsBranchTarget (MethodDefinition method, Instruction target)
+		{
+			foreach (var instr in method.Body.Instructions) {
+				if (IsBranch (instr) && instr.Operand == target)
+					return true;
+			}
+			return false;
+		}
+
+		static OpCode [] branches = new OpCode [] {
+			OpCodes.Br_S, OpCodes.Brfalse_S, OpCodes.Brtrue_S,
+			OpCodes.Beq_S, OpCodes.Bge_S, OpCodes.Bgt_S,
+			OpCodes.Ble_S, OpCodes.Blt_S, OpCodes.Bne_Un_S,
+			OpCodes.Bge_Un_S, OpCodes.Bgt_Un_S, OpCodes.Ble_Un_S,
+			OpCodes.Blt_Un_S, OpCodes.Br, OpCodes.Brfalse,
+			OpCodes.Brtrue, OpCodes.Beq, OpCodes.Bge,
+			OpCodes.Bgt, OpCodes.Ble, OpCodes.Blt,
+			OpCodes.Bne_Un, OpCodes.Bge_Un, OpCodes.Bgt_Un,
+			OpCodes.Ble_Un, OpCodes.Blt_Un
+		};
+
+		static bool IsBranch (Instruction instr)
+		{
+			return Array.IndexOf (branches, instr.OpCode) >= 0;
+		}
+
+		static TypeReference GetLastArgType (MethodDefinition method, Instruction instr)
+		{
+			var paramDef = GetOperandType(method, instr);
+			if (paramDef is null) {
+				throw new NotImplementedException ($"Last instruction before call to SetupBlockUnsafe ({instr.ToString ()}) was not a Ldfld, Ldarg or Ldloc - this is quite unexpected - something's changed in the code base!");
+			}
+			return paramDef;
+		}
+
+		static TypeReference? GetOperandType (MethodDefinition method, Instruction instr)
+		{
+			var thisOffset = method.IsStatic ? 0 : -1;
+			if (instr.OpCode == OpCodes.Ldarg_0)
+				return GetCheckParameterType (method, 0 + thisOffset);
+			if (instr.OpCode == OpCodes.Ldarg_1)
+				return GetCheckParameterType (method, 1 + thisOffset);
+			if (instr.OpCode == OpCodes.Ldarg_2)
+				return GetCheckParameterType (method, 2 + thisOffset);
+			if (instr.OpCode == OpCodes.Ldarg_3)
+				return GetCheckParameterType (method, 3 + thisOffset);
+			if (instr.OpCode == OpCodes.Ldarg_S)
+				return (instr.Operand as ParameterDefinition)?.ParameterType;
+			if (instr.OpCode == OpCodes.Ldloc_0)
+				return GetLocalType (method, 0);
+
+			if (instr.OpCode == OpCodes.Ldloc_1)
+				return GetLocalType (method, 1);
+
+			if (instr.OpCode == OpCodes.Ldloc_2)
+				return GetLocalType (method, 2);
+
+			if (instr.OpCode == OpCodes.Ldloc_3)
+				return GetLocalType (method, 3);
+
+			if (instr.OpCode == OpCodes.Ldloc_S)
+				return (instr.Operand as VariableDefinition)?.VariableType;
+			if (instr.OpCode == OpCodes.Ldsfld)
+				return (instr.Operand as FieldReference)?.FieldType;
+
+			return null;
+		}
+
+		static TypeReference GetCheckParameterType (MethodDefinition method, int index)
+		{
+			if (index >= method.Parameters.Count ()) {
```

Comment:
I'm not sure the reason for this CWL, since we don't return won't we crash on 275 with a out of bounds, and if so why have the print?

Reply:
Because I hit this and printing out exactly the method and the parameter tells me exactly what caused this. Out of bounds doesn't tell me this. I thought about pulling it out, but it's so much more helpful to have it there.

Reply:
If it's a failure case, consider just turning it into an assert.